### PR TITLE
fix: update web3py to v6, apply no-implicit-optional, misc fixes

### LINKED ISSUE #322: Error importing Uniswap with the latest version of web3.py
**Describe the bug**
```python
>>> from uniswap import Uniswap
from web3.contract import Contract, ContractFunction
ImportError: cannot import name 'ContractFunction' from 'web3.contract'
```

**To Reproduce**
```bash
$ python3.11 -m venv .venv
$ source .venv/bin/activate
(.venv) $ pip install git+https://github.com/ethereum/web3.py.git
(.venv) $ pip install uniswap-python
(.venv) $ python
Python 3.11.2 (main, Feb  8 2023, 14:49:24) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from uniswap import Uniswap
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/arjuna/.../review_uniswap_python/.venv/lib/python3.11/site-packages/uniswap/__init__.py", line 2, in <module>
    from .uniswap import Uniswap, _str_to_addr
  File "/home/arjuna/.../review_uniswap_python/.venv/lib/python3.11/site-packages/uniswap/uniswap.py", line 11, in <module>
    from web3.contract import Contract, ContractFunction
ImportError: cannot import name 'ContractFunction' from 'web3.contract' (/home/arjuna/.../review_uniswap_python/.venv/lib/python3.11/site-packages/web3/contract/__init__.py)
>>> from web3.contract import ContractFunction
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'ContractFunction' from 'web3.contract' (/home/arjuna/.../review_uniswap_python/.venv/lib/python3.11/site-packages/web3/contract/__init__.py)
>>> from web3.contract.contract import ContractFunction
>>>
```

**Expected behavior**
We should be able to import `Uniswap` with the latest version of web3.

**Additional context**
The reason I installed web3.py from GitHub repo is Python3.11 doesn't work with web3.py from pip.

```bash
ImportError: cannot import name 'getargspec' from 'inspect' (/usr/lib/python3.11/inspect.py)
```

### PR DESCRIPTION
**Pull request by ErikBjare**
Fixes #322

### CODE REVIEW

**Comment by ErikBjare** [comment]
I'm stumped, I keep getting `Revert reason: Not WETH9` for some reason :<

File: tests/test_uniswap.py [hunk 1]
```diff
@@ -53,7 +57,7 @@ def client(request, web3: Web3, ganache: GanacheInstance):
     )
 
 
-@pytest.fixture(scope="function", params=UNISWAP_VERSIONS)
+@pytest.fixture(scope="function")
```
Comment (ErikBjare) [inline tests/test_uniswap.py:60]: This was adding unnecessary complexity & number of tests.

File: uniswap/uniswap.py [hunk 2]
```diff
@@ -1435,12 +1435,12 @@ def _deadline(self) -> int:
...
 
     def _build_and_send_tx(
-        self, function: ContractFunction, tx_params: Optional[TxParams] = None
+        self, function: BaseContractFunction, tx_params: Optional[TxParams] = None
     ) -> HexBytes:
         """Build and send a transaction."""
         if not tx_params:
             tx_params = self._get_tx_params()
-        transaction = function.build_transaction(tx_params)
+        transaction = function._build_transaction(tx_params)
```
Comment (ErikBjare) [inline uniswap/uniswap.py:1443]: Maybe something changed here with the renaming?

File: uniswap/uniswap.py [hunk 3]
```diff
@@ -1474,8 +1476,25 @@ def _get_tx_params(self, value: Wei = Wei(0), gas: Wei = None) -> TxParams:
...
+        # if gasPrice still not set, use maxFeePerGas and maxPriorityFeePerGas
+        # TODO: Don't hard-code
+        if "gasPrice" not in params:
+            params["maxFeePerGas"] = Wei(200 * 1_000_000_000)  # 200 gwei
+            params["maxPriorityFeePerGas"] = Wei(self.w3.eth.max_priority_fee)
+
+        # Set chainId if not set
+        if "chainId" not in params:
+            params["chainId"] = self.netid
+
```
Comment (ErikBjare) [inline uniswap/uniswap.py:1497]: Idk why I need to set all this stuff now :/

File: uniswap/util.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 def _get_eth_simple_cache_middleware() -> Middleware:
     return construct_simple_cache_middleware(
-        cache_class=cast(Type[Dict[Any, Any]], functools.partial(lru.LRU, 256)),
+        cache=functools.partial(lru.LRU, 256),  # type: ignore
```
Comment (ErikBjare) [inline uniswap/util.py:21]: Is this correct? I seem to get a lot of chainId hits on my local ganache node...

File: uniswap/uniswap.py (hunk 2 above)
Comment (ErikBjare) [reply uniswap/uniswap.py]: Ugh, yeah this was it.

**Comment by ErikBjare** [comment]
Alright, finally done now. 

Shipping :shipit:

**Comment by liquid-8** [comment]
I couldn't manage to review it in time; sorry for that. I was going to take a look today. However, I'm glad it is fine now.

**Pull request merged**